lib/chore_runner/downloads/storage_service.ex

```diff
@@ -0,0 +1,62 @@
+defmodule ChoreRunner.Downloads.StorageService do
+  @typedoc """
+  Must have the appropriate file extension to be considered a valid name.
+  """
+  @type name :: String.t()
+  @type file_path :: String.t()
+  @type body :: binary()
+  @typedoc """
+  While both :path and :body are optional, at least one is required. The application will raise if both are provided.
+  """
+  @type opts :: [path: file_path(), body: body()]
+  @type reason :: any()
+  @type file :: %{
+          required(:id) => String.t(),
+          required(:name) => String.t(),
+          required(:type) => String.t(),
+          required(:created_at) => DateTime.t(),
+          chore_id: String.t()
+        }
+  @callback save_file(name, opts) :: {:ok, file()} | {:error, reason()}
+  @callback list_files() :: {:ok, [file()]} | {:error, reason()}
+  @callback delete_file(file()) :: :ok | {:error, reason()}
+  @callback file_url(file(), list()) :: String.t()
+  # @callback extra_ui_info() :: map()
+
+  @storage_service Application.get_env(
+                     :chore_runner,
+                     :storage_service,
+                     ChoreRunner.Downloads.TemporaryDiskStorageService
+                   )
+
+  def save_file(name, opts) do
+    cond do
+      is_nil(Path.extname(name)) ->
```

Comment:
Dialyzer gives an error that this can't match. I don't think `Path.extname/1` can return anything other than a binary

Reply:
Yeah turns out if a file doesn't have an extension its still valid